proposals/custom-element-default-focus.md

```diff
@@ -0,0 +1,70 @@
+
+# Default focus behaviors for custom elements
+
+Author: @rakina, @domenic, @tkent-google
+
+## Introduction
+
+Built-in elements are treated in different ways depending on their tag, attribute, style, etc. in terms of focus. Some built-in elements need an explicit `tabindex` set to be focusable, some can be focused through calling `focus()`, some others can be focused with clicking or through the tab key focus navigation, etc. In some cases their focusability also depends on the user agent and the OS, where certain preferences can make certain types of built-in elements to be focusable/not focusable. For example in macOS, editable form controls are click-focusable while non-editable form controls are not click-focusable.
+
+Meanwhile, all custom elements are treated the same for focus: they aren’t focusable by default, and they can only be made focusable by explicitly setting the `tabindex` attribute. This causes problems because the custom element user might also manipulate the `tabindex` attribute, causing a loss of default when the user changes the value and then removes it - leaving the custom element without any `tabindex` value.
+
+### Goals
+-   Give a way for custom elements to have the same range of focus behaviours as built-in elements.
+    
+### Non-goals
+-   Give a way to change the default focus behavior for built-in elements.
+-   Give a way for custom elements to have fully custom focus behavior that doesn’t follow a certain built-in element.
+    
+
+## Example
+
+```html
+<script>
+class MyButton extends HTMLElement {
+  constructor()  {
+    super();
+    this._internals =  this.attachInternals();
+    this._internals.matchFocusBehaviorOf =  "input[type='button']";
+  }
+}
+</script>
+<!-- these two buttons would behave the same for focus -->
+<input type="button" value="Click me">
+<my-button>Click me</my-button>
+```
+
+## Proposal
+
+Add a new property `matchFocusBehaviorOf` to `ElementInternals`, which accepts a string containing a tagname and optionally an attribute with a value (so it will accept values like `div`, `a[href]`, `input[type='checkbox']` , but not `.class`, `*`, etc.). The custom element will then be treated as if it has that tagname and attribute in terms of focus.
+
+## Alternatives considered
+
+### Using other ways to represent the built-in elements
+
+The proposed API above uses a CSS selectors to represent the built-in elements, but we’ve also considered some variants:
+
+ - Using “tagname + one attribute name + attribute’s value” combination,
+   e.g. `elementInternals.matchFocusBehaviorOf = {"input", "type," "button"}`
```

Comment:
This isn't valid JS syntax. Maybe `matchFocusBehaviorOf("input", "type", "button")` or three separate attributes or something like that.

Reply:
Oops, changed. Thanks!